Fix for some edge-case scenarios where ownerDocument is null.

Darkseal opened the pull request:
ownerDocument isn't guaranteed to have value. If you do something like $(document).tooltip(), the following error will occur:

Uncaught TypeError: Cannot read property ‘documentElement’ of null

This additional check (with a safe fallback provided) will fix it.

Darkseal committed:
Fix for some edge-case scenarios where ownerDocument is null.

ownerDocument isn't guaranteed to have value. If you do something like $(document).tooltip(), the following error will occur:

Uncaught TypeError: Cannot read property ‘documentElement’ of null

This additional check (with a safe fallback provided) will fix it.

twbs-savage commented:
**Tests passed.** Automated cross-browser testing via Sauce Labs and Travis CI shows that the JavaScript changes in this pull request are: **CONFIRMED**

Commit: 70db566f88c15c831c6a28b81c6838a8cac520f1
Build details: https://travis-ci.org/twbs-savage/bootstrap/builds/45790071

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

cvrebert commented:
Please add a relevant testcase to https://github.com/twbs/bootstrap/blob/master/js/tests/unit/tooltip.js

Darkseal committed:
Unit test added

Added a new unit test to tooltip.js to demonstrate the fix to the
document-attaching bug.

Darkseal commented:
Done. You can check the new test against the current 3.3.1 code and my code.

Darkseal committed:
Fixed some whitespaces

cvrebert commented:
So, what's your use-case for doing `$(document).tooltip()` (without the `selector` option) in the first place?

twbs-savage commented:
**Tests passed.** Automated cross-browser testing via Sauce Labs and Travis CI shows that the JavaScript changes in this pull request are: **CONFIRMED**

Commit: b0c622775e9a4ee8a62d3685883742c47f34d11d
Build details: https://travis-ci.org/twbs-savage/bootstrap/builds/45846886

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

Darkseal commented:
You can see my comment on #14483: cases might vary from 'JQueryUI/Tooltip' migrations to missing 'selector' option. I felt like the library should treat any element outside the DOM in the same exact way: gracefully exiting without showing the popup. Just like the 'html' selector, for example, which also falls into this same scenario and was already handled this way. 

'$("html").tooltip()' -> no errors, no tooltip (in 3.3.1)
'$(document).tooltip()' -> show errors (in 3.3.1)

That's what I wanted to normalize/fix.

Darkseal committed:
Fixed invalid quote mark as per Travis CI report.

twbs-savage commented:
**Tests passed.** Automated cross-browser testing via Sauce Labs and Travis CI shows that the JavaScript changes in this pull request are: **CONFIRMED**

Commit: 6a2fcc454c789e3c112ebc8b9541c22dc9ca73b5
Build details: https://travis-ci.org/twbs-savage/bootstrap/builds/45848141

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

cvrebert commented:
What about making `$(document).tooltip()`-without-a-`selector` instead immediately throw a meaningful error?

Darkseal commented:
I could work on that's what you want. It's also worth noticing that currently the `inDom` check (tooltip.js, line 155) isn't really doing what it states, since it evaluates `true` only if the element is contained in the document's `html` element: it should probably be renamed to something like `inDocumentElement` or something like that for better clarity. 

Once we do that, we could just replace that opaque `return` statement (if `inDocumentElement` is `false`) with a meaningful error describing the wrong user configuration issue.

Let me know if you want me to work on this, I love your library and I'll gladly help.

cvrebert commented:
Closing in favor of #15496.
